// @nats-io/jetstream/lib/jsapi_types.js
"use strict";
/*
 * Copyright 2023-2024 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PubHeaders = exports.PriorityPolicy = exports.ConsumerApiAction = exports.StoreCompression = exports.ReplayPolicy = exports.AckPolicy = exports.DeliverPolicy = exports.StorageType = exports.DiscardPolicy = exports.RetentionPolicy = void 0;
exports.defaultConsumer = defaultConsumer;
const nats_core_1 = require("@nats-io/nats-core");
exports.RetentionPolicy = {
    /**
     * Retain messages until the limits are reached, then trigger the discard policy.
     */
    Limits: "limits",
    /**
     * Retain messages while there is consumer interest on the particular subject.
     */
    Interest: "interest",
    /**
     * Retain messages until acknowledged
     */
    Workqueue: "workqueue",
};
exports.DiscardPolicy = {
    /**
     * Discard old messages to make room for the new ones
     */
    Old: "old",
    /**
     * Discard the new messages
     */
    New: "new",
};
exports.StorageType = {
    /**
     * Store persistently on files
     */
    File: "file",
    /**
     * Store in server memory - doesn't survive server restarts
     */
    Memory: "memory",
};
exports.DeliverPolicy = {
    /**
     * Deliver all messages
     */
    All: "all",
    /**
     * Deliver starting with the last message
     */
    Last: "last",
    /**
     * Deliver starting with new messages
     */
    New: "new",
    /**
     * Deliver starting with the specified sequence
     */
    StartSequence: "by_start_sequence",
    /**
     * Deliver starting with the specified time
     */
    StartTime: "by_start_time",
    /**
     * Deliver starting with the last messages for every subject
     */
    LastPerSubject: "last_per_subject",
};
exports.AckPolicy = {
    /**
     * Messages don't need to be Ack'ed.
     */
    None: "none",
    /**
     * Ack, acknowledges all messages with a lower sequence
     */
    All: "all",
    /**
     * All sequences must be explicitly acknowledged
     */
    Explicit: "explicit",
    /**
     * @ignore
     */
    NotSet: "",
};
exports.ReplayPolicy = {
    /**
     * Replays messages as fast as possible
     */
    Instant: "instant",
    /**
     * Replays messages following the original delay between messages
     */
    Original: "original",
};
exports.StoreCompression = {
    /**
     * No compression
     */
    None: "none",
    /**
     * S2 compression
     */
    S2: "s2",
};
exports.ConsumerApiAction = {
    CreateOrUpdate: "",
    Update: "update",
    Create: "create",
};
exports.PriorityPolicy = {
    None: "none",
    Overflow: "overflow",
    PinnedClient: "pinned_client",
};
function defaultConsumer(name, opts = {}) {
    return Object.assign({
        name: name,
        deliver_policy: exports.DeliverPolicy.All,
        ack_policy: exports.AckPolicy.Explicit,
        ack_wait: (0, nats_core_1.nanos)(30 * 1000),
        replay_policy: exports.ReplayPolicy.Instant,
    }, opts);
}
exports.PubHeaders = {
    MsgIdHdr: "Nats-Msg-Id",
    ExpectedStreamHdr: "Nats-Expected-Stream",
    ExpectedLastSeqHdr: "Nats-Expected-Last-Sequence",
    ExpectedLastMsgIdHdr: "Nats-Expected-Last-Msg-Id",
    ExpectedLastSubjectSequenceHdr: "Nats-Expected-Last-Subject-Sequence",
    ExpectedLastSubjectSequenceSubjectHdr: "Nats-Expected-Last-Subject-Sequence-Subject",
    /**
     * Sets the TTL for a message (Nanos value). Only have effect on streams that
     * enable {@link StreamConfig#allow_msg_ttl}.
     */
    MessageTTL: "Nats-TTL",
};
