// @nats-io/jetstream/lib/jsbaseclient_api.js
"use strict";
/*
 * Copyright 2021-2023 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseApiClientImpl = void 0;
exports.defaultJsOptions = defaultJsOptions;
const internal_1 = require("@nats-io/nats-core/internal");
const jserrors_1 = require("./jserrors");
const defaultPrefix = "$JS.API";
const defaultTimeout = 5000;
function defaultJsOptions(opts) {
    opts = opts || {};
    if (opts.domain) {
        opts.apiPrefix = `$JS.${opts.domain}.API`;
        delete opts.domain;
    }
    return (0, internal_1.extend)({ apiPrefix: defaultPrefix, timeout: defaultTimeout }, opts);
}
class BaseApiClientImpl {
    nc;
    opts;
    prefix;
    timeout;
    constructor(nc, opts) {
        this.nc = nc;
        this.opts = defaultJsOptions(opts);
        this._parseOpts();
        this.prefix = this.opts.apiPrefix;
        this.timeout = this.opts.timeout;
    }
    getOptions() {
        return Object.assign({}, this.opts);
    }
    _parseOpts() {
        let prefix = this.opts.apiPrefix;
        if (!prefix || prefix.length === 0) {
            throw internal_1.errors.InvalidArgumentError.format("prefix", "cannot be empty");
        }
        const c = prefix[prefix.length - 1];
        if (c === ".") {
            prefix = prefix.substr(0, prefix.length - 1);
        }
        this.opts.apiPrefix = prefix;
    }
    async _request(subj, data = null, opts) {
        opts = opts || {};
        opts.timeout = this.timeout;
        let a = internal_1.Empty;
        if (data) {
            a = new TextEncoder().encode(JSON.stringify(data));
        }
        let { retries } = opts;
        retries = retries || 1;
        retries = retries === -1 ? Number.MAX_SAFE_INTEGER : retries;
        const bo = (0, internal_1.backoff)();
        for (let i = 0; i < retries; i++) {
            try {
                const m = await this.nc.request(subj, a, opts);
                return this.parseJsResponse(m);
            }
            catch (err) {
                const re = err instanceof internal_1.RequestError ? err : null;
                if ((err instanceof internal_1.errors.TimeoutError || re?.isNoResponders()) &&
                    i + 1 < retries) {
                    await (0, internal_1.delay)(bo.backoff(i));
                }
                else {
                    throw re?.isNoResponders()
                        ? new jserrors_1.JetStreamNotEnabled("jetstream is not enabled", {
                            cause: err,
                        })
                        : err;
                }
            }
        }
    }
    async findStream(subject) {
        const q = { subject };
        const r = await this._request(`${this.prefix}.STREAM.NAMES`, q);
        const names = r;
        if (!names.streams || names.streams.length !== 1) {
            throw jserrors_1.StreamNotFoundError.fromMessage("no stream matches subject");
        }
        return names.streams[0];
    }
    getConnection() {
        return this.nc;
    }
    parseJsResponse(m) {
        const v = JSON.parse(new TextDecoder().decode(m.data));
        const r = v;
        if (r.error) {
            switch (r.error.err_code) {
                case jserrors_1.JetStreamApiCodes.ConsumerNotFound:
                    throw new jserrors_1.ConsumerNotFoundError(r.error);
                case jserrors_1.JetStreamApiCodes.StreamNotFound:
                    throw new jserrors_1.StreamNotFoundError(r.error);
                case jserrors_1.JetStreamApiCodes.JetStreamNotEnabledForAccount: {
                    const jserr = new jserrors_1.JetStreamApiError(r.error);
                    throw new jserrors_1.JetStreamNotEnabled(jserr.message, { cause: jserr });
                }
                default:
                    throw new jserrors_1.JetStreamApiError(r.error);
            }
        }
        return v;
    }
}
exports.BaseApiClientImpl = BaseApiClientImpl;
