// @nats-io/jetstream/lib/jsclient.js
"use strict";
/*
 * Copyright 2022-2024 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.JetStreamClientImpl = exports.JetStreamManagerImpl = void 0;
exports.toJetStreamClient = toJetStreamClient;
exports.jetstream = jetstream;
exports.jetstreamManager = jetstreamManager;
const jsbaseclient_api_1 = require("./jsbaseclient_api");
const jsmconsumer_api_1 = require("./jsmconsumer_api");
const internal_1 = require("@nats-io/nats-core/internal");
const jsmstream_api_1 = require("./jsmstream_api");
const internal_2 = require("@nats-io/nats-core/internal");
const jsapi_types_1 = require("./jsapi_types");
const jserrors_1 = require("./jserrors");
const jsm_direct_1 = require("./jsm_direct");
function toJetStreamClient(nc) {
    //@ts-ignore: see if we have a nc
    if (typeof nc.nc === "undefined") {
        return jetstream(nc);
    }
    return nc;
}
/**
 * Returns a {@link JetStreamClient} supported by the specified NatsConnection
 * @param nc
 * @param opts
 */
function jetstream(nc, opts = {}) {
    return new JetStreamClientImpl(nc, opts);
}
/**
 * Returns a {@link JetStreamManager} supported by the specified NatsConnection
 * @param nc
 * @param opts
 */
async function jetstreamManager(nc, opts = {}) {
    const adm = new JetStreamManagerImpl(nc, opts);
    if (opts.checkAPI !== false) {
        try {
            await adm.getAccountInfo();
        }
        catch (err) {
            throw err;
        }
    }
    return adm;
}
class JetStreamManagerImpl extends jsbaseclient_api_1.BaseApiClientImpl {
    streams;
    consumers;
    direct;
    constructor(nc, opts) {
        super(nc, opts);
        this.streams = new jsmstream_api_1.StreamAPIImpl(nc, opts);
        this.consumers = new jsmconsumer_api_1.ConsumerAPIImpl(nc, opts);
        this.direct = new jsm_direct_1.DirectStreamAPIImpl(nc, opts);
    }
    async getAccountInfo() {
        const r = await this._request(`${this.prefix}.INFO`);
        return r;
    }
    jetstream() {
        return jetstream(this.nc, this.getOptions());
    }
    advisories() {
        const iter = new internal_1.QueuedIteratorImpl();
        this.nc.subscribe(`$JS.EVENT.ADVISORY.>`, {
            callback: (err, msg) => {
                if (err) {
                    throw err;
                }
                try {
                    const d = this.parseJsResponse(msg);
                    const chunks = d.type.split(".");
                    const kind = chunks[chunks.length - 1];
                    iter.push({ kind: kind, data: d });
                }
                catch (err) {
                    iter.stop(err);
                }
            },
        });
        return iter;
    }
}
exports.JetStreamManagerImpl = JetStreamManagerImpl;
class JetStreamClientImpl extends jsbaseclient_api_1.BaseApiClientImpl {
    consumers;
    streams;
    consumerAPI;
    streamAPI;
    constructor(nc, opts) {
        super(nc, opts);
        this.consumerAPI = new jsmconsumer_api_1.ConsumerAPIImpl(nc, opts);
        this.streamAPI = new jsmstream_api_1.StreamAPIImpl(nc, opts);
        this.consumers = new jsmstream_api_1.ConsumersImpl(this.consumerAPI);
        this.streams = new jsmstream_api_1.StreamsImpl(this.streamAPI);
    }
    jetstreamManager(checkAPI) {
        if (checkAPI === undefined) {
            checkAPI = this.opts.checkAPI;
        }
        const opts = Object.assign({}, this.opts, { checkAPI });
        return jetstreamManager(this.nc, opts);
    }
    get apiPrefix() {
        return this.prefix;
    }
    async publish(subj, data = internal_1.Empty, opts) {
        opts = opts || {};
        opts = { ...opts };
        opts.expect = opts.expect || {};
        const mh = opts?.headers || (0, internal_2.headers)();
        if (opts) {
            if (opts.msgID) {
                mh.set(jsapi_types_1.PubHeaders.MsgIdHdr, opts.msgID);
            }
            if (opts.expect.lastMsgID) {
                mh.set(jsapi_types_1.PubHeaders.ExpectedLastMsgIdHdr, opts.expect.lastMsgID);
            }
            if (opts.expect.streamName) {
                mh.set(jsapi_types_1.PubHeaders.ExpectedStreamHdr, opts.expect.streamName);
            }
            if (typeof opts.expect.lastSequence === "number") {
                mh.set(jsapi_types_1.PubHeaders.ExpectedLastSeqHdr, `${opts.expect.lastSequence}`);
            }
            if (typeof opts.expect.lastSubjectSequence === "number") {
                mh.set(jsapi_types_1.PubHeaders.ExpectedLastSubjectSequenceHdr, `${opts.expect.lastSubjectSequence}`);
            }
            if (opts.expect.lastSubjectSequenceSubject) {
                mh.set(jsapi_types_1.PubHeaders.ExpectedLastSubjectSequenceSubjectHdr, opts.expect.lastSubjectSequenceSubject);
            }
            if (opts.ttl) {
                mh.set(jsapi_types_1.PubHeaders.MessageTTL, `${opts.ttl}`);
            }
        }
        const to = opts.timeout || this.timeout;
        const ro = {};
        if (to) {
            ro.timeout = to;
        }
        if (opts) {
            ro.headers = mh;
        }
        let { retries } = opts;
        retries = retries || 1;
        const bo = (0, internal_1.backoff)();
        let r = null;
        for (let i = 0; i < retries; i++) {
            try {
                r = await this.nc.request(subj, data, ro);
                // if here we succeeded
                break;
            }
            catch (err) {
                const re = err instanceof internal_2.RequestError ? err : null;
                if ((err instanceof internal_2.errors.TimeoutError || re?.isNoResponders()) &&
                    i + 1 < retries) {
                    await (0, internal_1.delay)(bo.backoff(i));
                }
                else {
                    throw re?.isNoResponders()
                        ? new jserrors_1.JetStreamNotEnabled(`jetstream is not enabled`, {
                            cause: err,
                        })
                        : err;
                }
            }
        }
        const pa = this.parseJsResponse(r);
        if (pa.stream === "") {
            throw new jserrors_1.JetStreamError("invalid ack response");
        }
        pa.duplicate = pa.duplicate ? pa.duplicate : false;
        return pa;
    }
}
exports.JetStreamClientImpl = JetStreamClientImpl;
