// @nats-io/jetstream/lib/jserrors.js
"use strict";
/*
 * Copyright 2024 Synadia Communications, Inc
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.jserrors = exports.StreamNotFoundError = exports.ConsumerNotFoundError = exports.JetStreamApiError = exports.InvalidNameError = exports.JetStreamApiCodes = exports.JetStreamStatus = exports.JetStreamStatusError = exports.JetStreamError = exports.JetStreamNotEnabled = void 0;
exports.isMessageNotFound = isMessageNotFound;
const types_1 = require("./types");
class JetStreamNotEnabled extends Error {
    constructor(message, opts) {
        super(message, opts);
        this.name = "JetStreamNotEnabled";
    }
}
exports.JetStreamNotEnabled = JetStreamNotEnabled;
class JetStreamError extends Error {
    constructor(message, opts) {
        super(message, opts);
        this.name = "JetStreamError";
    }
}
exports.JetStreamError = JetStreamError;
class JetStreamStatusError extends JetStreamError {
    code;
    constructor(message, code, opts) {
        super(message, opts);
        this.code = code;
        this.name = "JetStreamStatusError";
    }
}
exports.JetStreamStatusError = JetStreamStatusError;
class JetStreamStatus {
    msg;
    _description;
    constructor(msg) {
        this.msg = msg;
        this._description = "";
    }
    static maybeParseStatus(msg) {
        const status = new JetStreamStatus(msg);
        return status.code === 0 ? null : status;
    }
    toError() {
        return new JetStreamStatusError(this.description, this.code);
    }
    debug() {
        console.log({
            subject: this.msg.subject,
            reply: this.msg.reply,
            description: this.description,
            status: this.code,
            headers: this.msg.headers,
        });
    }
    get code() {
        return this.msg.headers?.code || 0;
    }
    get description() {
        if (this._description === "") {
            this._description = this.msg.headers?.description?.toLowerCase() || "";
            if (this._description === "") {
                this._description = this.code === 503 ? "no responders" : "unknown";
            }
        }
        return this._description;
    }
    isIdleHeartbeat() {
        return this.code === 100 && this.description === "idle heartbeat";
    }
    isFlowControlRequest() {
        return this.code === 100 && this.description === "flowcontrol request";
    }
    parseHeartbeat() {
        if (this.isIdleHeartbeat()) {
            return {
                type: "heartbeat",
                lastConsumerSequence: parseInt(this.msg.headers?.get("Nats-Last-Consumer") || "0"),
                lastStreamSequence: parseInt(this.msg.headers?.get("Nats-Last-Stream") || "0"),
            };
        }
        return null;
    }
    isRequestTimeout() {
        return this.code === 408 && this.description === "request timeout";
    }
    parseDiscard() {
        const discard = {
            msgsLeft: 0,
            bytesLeft: 0,
        };
        const msgsLeft = this.msg.headers?.get(types_1.JsHeaders.PendingMessagesHdr);
        if (msgsLeft) {
            discard.msgsLeft = parseInt(msgsLeft);
        }
        const bytesLeft = this.msg.headers?.get(types_1.JsHeaders.PendingBytesHdr);
        if (bytesLeft) {
            discard.bytesLeft = parseInt(bytesLeft);
        }
        return discard;
    }
    isBadRequest() {
        return this.code === 400;
    }
    isConsumerDeleted() {
        return this.code === 409 && this.description === "consumer deleted";
    }
    isStreamDeleted() {
        return this.code === 409 && this.description === "stream deleted";
    }
    isIdleHeartbeatMissed() {
        return this.code === 409 && this.description === "idle heartbeats missed";
    }
    isMaxWaitingExceeded() {
        return this.code === 409 && this.description === "exceeded maxwaiting";
    }
    isConsumerIsPushBased() {
        return this.code === 409 && this.description === "consumer is push based";
    }
    isExceededMaxWaiting() {
        return this.code === 409 &&
            this.description.includes("exceeded maxwaiting");
    }
    isExceededMaxRequestBatch() {
        return this.code === 409 &&
            this.description.includes("exceeded maxrequestbatch");
    }
    isExceededMaxExpires() {
        return this.code === 409 &&
            this.description.includes("exceeded maxrequestexpires");
    }
    isExceededLimit() {
        return this.isExceededMaxExpires() || this.isExceededMaxWaiting() ||
            this.isExceededMaxRequestBatch() || this.isMessageSizeExceedsMaxBytes();
    }
    isMessageNotFound() {
        return this.code === 404 && this.description === "message not found";
    }
    isMessageSizeExceedsMaxBytes() {
        return this.code === 409 &&
            this.description === "message size exceeds maxbytes";
    }
    isEndOfBatch() {
        return this.code === 204 && this.description === "eob";
    }
}
exports.JetStreamStatus = JetStreamStatus;
exports.JetStreamApiCodes = {
    ConsumerNotFound: 10014,
    StreamNotFound: 10059,
    JetStreamNotEnabledForAccount: 10039,
    StreamWrongLastSequence: 10071,
    NoMessageFound: 10037,
};
function isMessageNotFound(err) {
    return err instanceof JetStreamApiError &&
        err.code === exports.JetStreamApiCodes.NoMessageFound;
}
class InvalidNameError extends Error {
    constructor(message = "", opts) {
        super(message, opts);
        this.name = "InvalidNameError";
    }
}
exports.InvalidNameError = InvalidNameError;
class JetStreamApiError extends Error {
    #apiError;
    constructor(jsErr, opts) {
        super(jsErr.description, opts);
        this.#apiError = jsErr;
        this.name = "JetStreamApiError";
    }
    get code() {
        return this.#apiError.err_code;
    }
    get status() {
        return this.#apiError.code;
    }
    apiError() {
        return Object.assign({}, this.#apiError);
    }
}
exports.JetStreamApiError = JetStreamApiError;
class ConsumerNotFoundError extends JetStreamApiError {
    constructor(jsErr, opts) {
        super(jsErr, opts);
        this.name = "ConsumerNotFoundError";
    }
}
exports.ConsumerNotFoundError = ConsumerNotFoundError;
class StreamNotFoundError extends JetStreamApiError {
    constructor(jsErr, opts) {
        super(jsErr, opts);
        this.name = "StreamNotFoundError";
    }
    static fromMessage(message) {
        return new StreamNotFoundError({
            err_code: exports.JetStreamApiCodes.StreamNotFound,
            description: message,
            code: 404,
        });
    }
}
exports.StreamNotFoundError = StreamNotFoundError;
exports.jserrors = {
    InvalidNameError,
    ConsumerNotFoundError,
    StreamNotFoundError,
    JetStreamError,
    JetStreamApiError,
    JetStreamNotEnabled,
};
