// @nats-io/jetstream/lib/jslister.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListerImpl = void 0;
/*
 * Copyright 2021-2024 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const internal_1 = require("@nats-io/nats-core/internal");
class ListerImpl {
    err;
    offset;
    pageInfo;
    subject;
    jsm;
    filter;
    payload;
    constructor(subject, filter, jsm, payload) {
        if (!subject) {
            throw internal_1.errors.InvalidArgumentError.format("subject", "is required");
        }
        this.subject = subject;
        this.jsm = jsm;
        this.offset = 0;
        this.pageInfo = {};
        this.filter = filter;
        this.payload = payload || {};
    }
    async next() {
        if (this.err) {
            return [];
        }
        if (this.pageInfo && this.offset >= this.pageInfo.total) {
            return [];
        }
        const offset = { offset: this.offset };
        if (this.payload) {
            Object.assign(offset, this.payload);
        }
        try {
            const r = await this.jsm._request(this.subject, offset, { timeout: this.jsm.timeout });
            this.pageInfo = r;
            // offsets are reported in total, so need to count
            // all the entries returned
            const count = this.countResponse(r);
            if (count === 0) {
                // we are done if we get a null set of infos
                return [];
            }
            this.offset += count;
            return this.filter(r);
        }
        catch (err) {
            this.err = err;
            throw err;
        }
    }
    countResponse(r) {
        switch (r?.type) {
            case "io.nats.jetstream.api.v1.stream_names_response":
            case "io.nats.jetstream.api.v1.stream_list_response":
                return r.streams?.length || 0;
            case "io.nats.jetstream.api.v1.consumer_list_response":
                return r.consumers?.length || 0;
            default:
                console.error(`jslister.ts: unknown API response for paged output: ${r?.type}`);
                // has to be a stream...
                return r.streams?.length || 0;
        }
    }
    async *[Symbol.asyncIterator]() {
        let page = await this.next();
        while (page.length > 0) {
            for (const item of page) {
                yield item;
            }
            page = await this.next();
        }
    }
}
exports.ListerImpl = ListerImpl;
