// @nats-io/jetstream/lib/jsm_direct.js
"use strict";
/*
 * Copyright 2021-2023 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DirectConsumer = exports.DirectMsgImpl = exports.DirectStreamAPIImpl = void 0;
const jsbaseclient_api_1 = require("./jsbaseclient_api");
const types_1 = require("./types");
const internal_1 = require("@nats-io/nats-core/internal");
const jsutil_1 = require("./jsutil");
const jserrors_1 = require("./jserrors");
class DirectStreamAPIImpl extends jsbaseclient_api_1.BaseApiClientImpl {
    constructor(nc, opts) {
        super(nc, opts);
    }
    async getMessage(stream, query) {
        (0, jsutil_1.validateStreamName)(stream);
        if ("start_time" in query) {
            const { min, ok } = this.nc.features.get(internal_1.Feature.JS_BATCH_DIRECT_GET);
            if (!ok) {
                throw new Error(`start_time direct option require server ${min}`);
            }
        }
        // if doing a last_by_subj request, we append the subject
        // this allows last_by_subj to be subject to permissions (KV)
        let qq = query;
        const { last_by_subj } = qq;
        if (last_by_subj) {
            qq = null;
        }
        const payload = qq ? JSON.stringify(qq) : internal_1.Empty;
        const pre = this.opts.apiPrefix || "$JS.API";
        const subj = last_by_subj
            ? `${pre}.DIRECT.GET.${stream}.${last_by_subj}`
            : `${pre}.DIRECT.GET.${stream}`;
        const r = await this.nc.request(subj, payload, { timeout: this.timeout });
        if (r.headers?.code !== 0) {
            const status = new jserrors_1.JetStreamStatus(r);
            if (status.isMessageNotFound()) {
                return Promise.resolve(null);
            }
            else {
                return Promise.reject(status.toError());
            }
        }
        const dm = new DirectMsgImpl(r);
        return Promise.resolve(dm);
    }
    getBatch(stream, opts) {
        opts.batch = opts.batch || 1024;
        return this.get(stream, opts);
    }
    getLastMessagesFor(stream, opts) {
        return this.get(stream, opts);
    }
    get(stream, opts) {
        opts = { ...opts };
        const { min, ok } = this.nc.features.get(internal_1.Feature.JS_BATCH_DIRECT_GET);
        if (!ok) {
            return Promise.reject(new Error(`batch direct require server ${min}`));
        }
        (0, jsutil_1.validateStreamName)(stream);
        const callback = typeof opts.callback === "function" ? opts.callback : null;
        const iter = new internal_1.QueuedIteratorImpl();
        function pushIter(done, d) {
            if (done) {
                iter.push(() => {
                    done.err ? iter.stop(done.err) : iter.stop();
                });
                return;
            }
            iter.push(d);
        }
        function pushCb(done, m) {
            const cb = callback;
            if (typeof m === "function") {
                m();
                return;
            }
            cb(done, m);
        }
        if (callback) {
            iter.iterClosed.then((err) => {
                push({ err: err ? err : undefined }, {});
                sub.unsubscribe();
            });
        }
        const push = callback ? pushCb : pushIter;
        const inbox = (0, internal_1.createInbox)(this.nc.options.inboxPrefix);
        let batchSupported = false;
        const sub = this.nc.subscribe(inbox, {
            timeout: 5000,
            callback: (err, msg) => {
                if (err) {
                    iter.stop(err);
                    sub.unsubscribe();
                    return;
                }
                const status = jserrors_1.JetStreamStatus.maybeParseStatus(msg);
                if (status) {
                    if (status.isEndOfBatch()) {
                        push({}, () => {
                            iter.stop();
                        });
                    }
                    else {
                        const err = status.toError();
                        push({ err }, () => {
                            iter.stop(err);
                        });
                    }
                    return;
                }
                if (!batchSupported) {
                    if (typeof msg.headers?.get("Nats-Num-Pending") !== "string") {
                        // no batch/max_bytes option was provided, so single response
                        sub.unsubscribe();
                        push({}, () => {
                            iter.stop();
                        });
                    }
                    else {
                        batchSupported = true;
                    }
                }
                push(null, new DirectMsgImpl(msg));
            },
        });
        const pre = this.opts.apiPrefix || "$JS.API";
        const subj = `${pre}.DIRECT.GET.${stream}`;
        const payload = JSON.stringify(opts, (key, value) => {
            if ((key === "up_to_time" || key === "start_time") && value instanceof Date) {
                return value.toISOString();
            }
            return value;
        });
        this.nc.publish(subj, payload, { reply: inbox });
        return Promise.resolve(iter);
    }
}
exports.DirectStreamAPIImpl = DirectStreamAPIImpl;
class DirectMsgImpl {
    data;
    header;
    constructor(m) {
        if (!m.headers) {
            throw new Error("headers expected");
        }
        this.data = m.data;
        this.header = m.headers;
    }
    get subject() {
        return this.header.last(types_1.DirectMsgHeaders.Subject);
    }
    get seq() {
        const v = this.header.last(types_1.DirectMsgHeaders.Sequence);
        return typeof v === "string" ? parseInt(v) : 0;
    }
    get time() {
        return new Date(Date.parse(this.timestamp));
    }
    get timestamp() {
        return this.header.last(types_1.DirectMsgHeaders.TimeStamp);
    }
    get stream() {
        return this.header.last(types_1.DirectMsgHeaders.Stream);
    }
    get lastSequence() {
        const v = this.header.last(types_1.DirectMsgHeaders.LastSequence);
        return typeof v === "string" ? parseInt(v) : 0;
    }
    get pending() {
        const v = this.header.last(types_1.DirectMsgHeaders.NumPending);
        // if we have a pending - this pending will include the number of messages
        // in the stream + the end of batch signal - better to remove the eob signal
        // from the count so the client can estimate how many messages are
        // in the stream. If a batch is 1 message, the pending is not included.
        return typeof v === "string" ? parseInt(v) - 1 : -1;
    }
    json(reviver) {
        return JSON.parse(new TextDecoder().decode(this.data), reviver);
    }
    string() {
        return internal_1.TD.decode(this.data);
    }
}
exports.DirectMsgImpl = DirectMsgImpl;
function isDirectBatchStartTime(t) {
    return typeof t === "object" && "start_time" in t;
}
function isMaxBytes(t) {
    return typeof t === "object" && "max_bytes" in t;
}
class DirectConsumer {
    stream;
    api;
    cursor;
    listeners;
    start;
    constructor(stream, api, start) {
        this.stream = stream;
        this.api = api;
        this.cursor = { last: 0 };
        this.listeners = [];
        this.start = start;
    }
    getOptions(opts) {
        opts = opts || {};
        const dbo = {};
        if (this.cursor.last === 0) {
            // we have never pulled, honor initial request options
            if (isDirectBatchStartTime(this.start)) {
                dbo.start_time = this.start.start_time;
            }
            else {
                dbo.seq = this.start.seq || 1;
            }
        }
        else {
            dbo.seq = this.cursor.last + 1;
        }
        if (isMaxBytes(opts)) {
            dbo.max_bytes = opts.max_bytes;
        }
        else {
            dbo.batch = opts.batch ?? 100;
        }
        return dbo;
    }
    status() {
        const iter = new internal_1.QueuedIteratorImpl();
        this.listeners.push(iter);
        return iter;
    }
    notify(n) {
        if (this.listeners.length > 0) {
            (() => {
                const remove = [];
                this.listeners.forEach((l) => {
                    const qi = l;
                    if (!qi.done) {
                        qi.push(n);
                    }
                    else {
                        remove.push(qi);
                    }
                });
                this.listeners = this.listeners.filter((l) => !remove.includes(l));
            })();
        }
    }
    debug() {
        console.log(this.cursor);
    }
    consume(opts) {
        let pending;
        let requestDone;
        const qi = new internal_1.QueuedIteratorImpl();
        (async () => {
            while (true) {
                // if we have nothing pending, slow it down
                // on the first pull pending doesn't exist so no delay
                if (this.cursor.pending === 0) {
                    this.notify({
                        type: "debug",
                        code: 0,
                        description: "sleeping for 2500",
                    });
                    pending = (0, internal_1.delay)(2500);
                    await pending;
                }
                // check that we are still supposed to be running
                // pending could have released if the iter closed
                if (qi.done) {
                    break;
                }
                requestDone = (0, internal_1.deferred)();
                const dbo = this.getOptions(opts);
                this.notify({
                    type: "next",
                    options: Object.assign({}, opts),
                });
                dbo.callback = (r, sm) => {
                    if (r) {
                        // if the current fetch is done, ready to schedule the next
                        if (r.err) {
                            if (r.err instanceof jserrors_1.JetStreamStatusError) {
                                this.notify({
                                    type: "debug",
                                    code: r.err.code,
                                    description: r.err.message,
                                });
                            }
                            else {
                                this.notify({
                                    type: "debug",
                                    code: 0,
                                    description: r.err.message,
                                });
                            }
                        }
                        requestDone.resolve();
                    }
                    else if (sm.lastSequence > 0 && sm.lastSequence !== this.cursor.last) {
                        // need to reset
                        src.stop();
                        requestDone.resolve();
                        this.notify({
                            type: "reset",
                            name: "direct",
                        });
                    }
                    else {
                        qi.push(sm);
                        qi.received++;
                        this.cursor.last = sm.seq;
                        this.cursor.pending = sm.pending;
                    }
                };
                const src = await this.api.getBatch(this.stream, dbo);
                qi.iterClosed.then(() => {
                    src.stop();
                    pending?.cancel();
                    requestDone?.resolve();
                });
                await requestDone;
            }
        })().catch((err) => {
            qi.stop(err);
        });
        return Promise.resolve(qi);
    }
    async fetch(opts) {
        const dbo = this.getOptions(opts);
        const qi = new internal_1.QueuedIteratorImpl();
        const src = await this.api.get(this.stream, Object.assign({
            callback: (done, sm) => {
                if (done) {
                    // the server sent error or is done, we are done
                    qi.push(() => {
                        done.err ? qi.stop(done.err) : qi.stop();
                    });
                }
                else if (sm.lastSequence > 0 && sm.lastSequence !== this.cursor.last) {
                    // we are done early because the sequence jumped unexpectedly
                    qi.push(() => {
                        qi.stop();
                    });
                    src.stop();
                }
                else {
                    // pass-through to client, and record
                    qi.push(sm);
                    qi.received++;
                    this.cursor.last = sm.seq;
                    this.cursor.pending = sm.pending;
                }
            },
        }, dbo));
        qi.iterClosed.then(() => {
            src.stop();
        });
        return qi;
    }
    async next() {
        const sm = await this.api.getMessage(this.stream, {
            seq: this.cursor.last + 1,
        });
        const seq = sm?.seq;
        if (seq) {
            this.cursor.last = seq;
        }
        return sm;
    }
}
exports.DirectConsumer = DirectConsumer;
