// @nats-io/jetstream/lib/jsmconsumer_api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsumerAPIImpl = void 0;
/*
 * Copyright 2021-2024 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const jsbaseclient_api_1 = require("./jsbaseclient_api");
const jslister_1 = require("./jslister");
const jsutil_1 = require("./jsutil");
const internal_1 = require("@nats-io/nats-core/internal");
const jsapi_types_1 = require("./jsapi_types");
class ConsumerAPIImpl extends jsbaseclient_api_1.BaseApiClientImpl {
    constructor(nc, opts) {
        super(nc, opts);
    }
    async addUpdate(stream, cfg, opts) {
        (0, jsutil_1.validateStreamName)(stream);
        opts = opts || {};
        if (cfg.deliver_group && cfg.flow_control) {
            throw internal_1.InvalidArgumentError.format(["flow_control", "deliver_group"], "are mutually exclusive");
        }
        if (cfg.deliver_group && cfg.idle_heartbeat) {
            throw internal_1.InvalidArgumentError.format(["idle_heartbeat", "deliver_group"], "are mutually exclusive");
        }
        if (isPriorityGroup(cfg)) {
            const { min, ok } = this.nc.features.get(internal_1.Feature.JS_PRIORITY_GROUPS);
            if (!ok) {
                throw new Error(`priority_groups require server ${min}`);
            }
            if (cfg.deliver_subject) {
                throw internal_1.InvalidArgumentError.format("deliver_subject", "cannot be set when using priority groups");
            }
            validatePriorityGroups(cfg);
        }
        const cr = {};
        cr.config = cfg;
        cr.stream_name = stream;
        cr.action = opts.action || jsapi_types_1.ConsumerApiAction.Create;
        cr.pedantic = opts.pedantic || false;
        if (cr.config.durable_name) {
            (0, jsutil_1.validateDurableName)(cr.config.durable_name);
        }
        const nci = this.nc;
        let { min, ok: newAPI } = nci.features.get(internal_1.Feature.JS_NEW_CONSUMER_CREATE_API);
        const name = cfg.name === "" ? undefined : cfg.name;
        if (name && !newAPI) {
            throw internal_1.InvalidArgumentError.format("name", `requires server ${min}`);
        }
        if (name) {
            try {
                (0, jsutil_1.minValidation)("name", name);
            }
            catch (err) {
                // if we have a cannot contain the message, massage a bit
                const m = err.message;
                const idx = m.indexOf("cannot contain");
                if (idx !== -1) {
                    throw new Error(`consumer 'name' ${m.substring(idx)}`);
                }
                throw err;
            }
        }
        let subj;
        let consumerName = "";
        // new api doesn't support multiple filter subjects
        // this delayed until here because the consumer in an update could have
        // been created with the new API, and have a `name`
        if (Array.isArray(cfg.filter_subjects)) {
            const { min, ok } = nci.features.get(internal_1.Feature.JS_MULTIPLE_CONSUMER_FILTER);
            if (!ok) {
                throw internal_1.InvalidArgumentError.format("filter_subjects", `requires server ${min}`);
            }
            newAPI = false;
        }
        if (cfg.metadata) {
            const { min, ok } = nci.features.get(internal_1.Feature.JS_STREAM_CONSUMER_METADATA);
            if (!ok) {
                throw internal_1.InvalidArgumentError.format("metadata", `requires server ${min}`);
            }
        }
        if (newAPI) {
            consumerName = cfg.name ?? cfg.durable_name ?? "";
        }
        if (consumerName !== "") {
            let fs = cfg.filter_subject ?? undefined;
            if (fs === ">") {
                fs = undefined;
            }
            subj = fs !== undefined
                ? `${this.prefix}.CONSUMER.CREATE.${stream}.${consumerName}.${fs}`
                : `${this.prefix}.CONSUMER.CREATE.${stream}.${consumerName}`;
        }
        else {
            subj = cfg.durable_name
                ? `${this.prefix}.CONSUMER.DURABLE.CREATE.${stream}.${cfg.durable_name}`
                : `${this.prefix}.CONSUMER.CREATE.${stream}`;
        }
        const r = await this._request(subj, cr);
        return r;
    }
    add(stream, cfg, opts) {
        // in the previous API 3rd arg was a hidden arg, the action - a string
        opts = opts || {};
        let action = jsapi_types_1.ConsumerApiAction.Create;
        if (typeof opts === "string") {
            action = opts;
            opts = {};
        }
        const cco = Object.assign({}, { action }, opts);
        return this.addUpdate(stream, cfg, cco);
    }
    async update(stream, durable, cfg) {
        const ci = await this.info(stream, durable);
        const changable = cfg;
        return this.addUpdate(stream, Object.assign(ci.config, changable), { action: jsapi_types_1.ConsumerApiAction.Update });
    }
    async info(stream, name) {
        (0, jsutil_1.validateStreamName)(stream);
        (0, jsutil_1.validateDurableName)(name);
        const r = await this._request(`${this.prefix}.CONSUMER.INFO.${stream}.${name}`);
        return r;
    }
    async delete(stream, name) {
        (0, jsutil_1.validateStreamName)(stream);
        (0, jsutil_1.validateDurableName)(name);
        const r = await this._request(`${this.prefix}.CONSUMER.DELETE.${stream}.${name}`);
        const cr = r;
        return cr.success;
    }
    list(stream) {
        (0, jsutil_1.validateStreamName)(stream);
        const filter = (v) => {
            const clr = v;
            return clr.consumers;
        };
        const subj = `${this.prefix}.CONSUMER.LIST.${stream}`;
        return new jslister_1.ListerImpl(subj, filter, this);
    }
    // Fixme: the API returns the number of nanoseconds, but really should return
    //  millis,
    pause(stream, name, until) {
        const subj = `${this.prefix}.CONSUMER.PAUSE.${stream}.${name}`;
        const opts = {
            pause_until: until.toISOString(),
        };
        return this._request(subj, opts);
    }
    // Fixme: the API returns the number of nanoseconds, but really should return
    //  millis,
    resume(stream, name) {
        return this.pause(stream, name, new Date(0));
    }
    unpin(stream, name, group) {
        const subj = `${this.prefix}.CONSUMER.UNPIN.${stream}.${name}`;
        return this._request(subj, { group });
    }
}
exports.ConsumerAPIImpl = ConsumerAPIImpl;
function isPriorityGroup(config) {
    const pg = config;
    return pg && pg.priority_groups !== undefined ||
        pg.priority_policy !== undefined;
}
function validatePriorityGroups(pg) {
    if (isPriorityGroup(pg)) {
        if (!Array.isArray(pg.priority_groups)) {
            throw internal_1.InvalidArgumentError.format(["priority_groups"], "must be an array");
        }
        if (pg.priority_groups.length === 0) {
            throw internal_1.InvalidArgumentError.format(["priority_groups"], "must have at least one group");
        }
        pg.priority_groups.forEach((g) => {
            (0, jsutil_1.minValidation)("priority_group", g);
            if (g.length > 16) {
                throw internal_1.errors.InvalidArgumentError.format("group", "must be 16 characters or less");
            }
        });
        if (pg.priority_policy !== jsapi_types_1.PriorityPolicy.None &&
            pg.priority_policy !== jsapi_types_1.PriorityPolicy.Overflow &&
            pg.priority_policy !== jsapi_types_1.PriorityPolicy.PinnedClient) {
            throw internal_1.InvalidArgumentError.format(["priority_policy"], "must be 'none', 'overflow', or 'pinned_client'");
        }
    }
}
