// @nats-io/jetstream/lib/jsmstream_api.js
"use strict";
/*
 * Copyright 2021-2025 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamsImpl = exports.StoredMsgImpl = exports.StreamAPIImpl = exports.StreamImpl = exports.ConsumersImpl = void 0;
exports.convertStreamSourceDomain = convertStreamSourceDomain;
const internal_1 = require("@nats-io/nats-core/internal");
const jsbaseclient_api_1 = require("./jsbaseclient_api");
const jslister_1 = require("./jslister");
const jsutil_1 = require("./jsutil");
const types_1 = require("./types");
const jsapi_types_1 = require("./jsapi_types");
const consumer_1 = require("./consumer");
const jsmconsumer_api_1 = require("./jsmconsumer_api");
const pushconsumer_1 = require("./pushconsumer");
const jserrors_1 = require("./jserrors");
function convertStreamSourceDomain(s) {
    if (s === undefined) {
        return undefined;
    }
    const { domain } = s;
    if (domain === undefined) {
        return s;
    }
    const copy = Object.assign({}, s);
    delete copy.domain;
    if (domain === "") {
        return copy;
    }
    if (copy.external) {
        throw internal_1.InvalidArgumentError.format(["domain", "external"], "are mutually exclusive");
    }
    copy.external = { api: `$JS.${domain}.API` };
    return copy;
}
class ConsumersImpl {
    api;
    notified;
    constructor(api) {
        this.api = api;
        this.notified = false;
    }
    checkVersion() {
        const fv = this.api.nc.features.get(internal_1.Feature.JS_SIMPLIFICATION);
        if (!fv.ok) {
            return Promise.reject(new Error(`consumers framework is only supported on servers ${fv.min} or better`));
        }
        return Promise.resolve();
    }
    async getPushConsumer(stream, name) {
        await this.checkVersion();
        (0, jsutil_1.minValidation)("stream", stream);
        if (typeof name === "string") {
            (0, jsutil_1.minValidation)("name", name);
            const ci = await this.api.info(stream, name);
            if (typeof ci.config.deliver_subject !== "string") {
                return Promise.reject(new Error("not a push consumer"));
            }
            return new pushconsumer_1.PushConsumerImpl(this.api, ci);
        }
        else if (name === undefined) {
            return this.getOrderedPushConsumer(stream);
        }
        else if ((0, types_1.isOrderedPushConsumerOptions)(name)) {
            const opts = name;
            return this.getOrderedPushConsumer(stream, opts);
        }
        return Promise.reject(new Error("unsupported push consumer type"));
    }
    async getOrderedPushConsumer(stream, opts = {}) {
        opts = Object.assign({}, opts);
        let { name_prefix, deliver_prefix, filter_subjects } = opts;
        delete opts.deliver_prefix;
        delete opts.name_prefix;
        delete opts.filter_subjects;
        if (typeof opts.opt_start_seq === "number") {
            opts.deliver_policy = jsapi_types_1.DeliverPolicy.StartSequence;
        }
        if (typeof opts.opt_start_time === "string") {
            opts.deliver_policy = jsapi_types_1.DeliverPolicy.StartTime;
        }
        name_prefix = name_prefix || `oc_${internal_1.nuid.next()}`;
        (0, jsutil_1.minValidation)("name_prefix", name_prefix);
        deliver_prefix = deliver_prefix ||
            (0, internal_1.createInbox)(this.api.nc.options.inboxPrefix);
        (0, jsutil_1.minValidation)("deliver_prefix", name_prefix);
        const cc = Object.assign({}, opts);
        cc.ack_policy = jsapi_types_1.AckPolicy.None;
        cc.inactive_threshold = (0, internal_1.nanos)(5 * 60 * 1000);
        cc.num_replicas = 1;
        cc.max_deliver = 1;
        cc.flow_control = true;
        cc.idle_heartbeat = (0, internal_1.nanos)(30_000);
        if (Array.isArray(filter_subjects)) {
            cc.filter_subjects = filter_subjects;
        }
        if (typeof filter_subjects === "string") {
            cc.filter_subject = filter_subjects;
        }
        if (typeof cc.filter_subjects === "undefined" &&
            typeof cc.filter_subject === "undefined") {
            cc.filter_subject = ">";
        }
        cc.name = `${name_prefix}_1`;
        cc.deliver_subject = `${deliver_prefix}.1`;
        const ci = await this.api.add(stream, cc);
        const iopts = {
            name_prefix,
            deliver_prefix,
            ordered: true,
        };
        return new pushconsumer_1.PushConsumerImpl(this.api, ci, iopts);
    }
    getBoundPushConsumer(opts) {
        if ((0, types_1.isBoundPushConsumerOptions)(opts)) {
            const ci = { config: opts };
            return Promise.resolve(new pushconsumer_1.PushConsumerImpl(this.api, ci, { bound: true }));
        }
        else {
            return Promise.reject(internal_1.errors.InvalidArgumentError.format("deliver_subject", "is required"));
        }
    }
    async get(stream, name) {
        await this.checkVersion();
        if (typeof name === "string") {
            const ci = await this.api.info(stream, name);
            if (typeof ci.config.deliver_subject === "string") {
                return Promise.reject(new Error("not a pull consumer"));
            }
            else {
                return new consumer_1.PullConsumerImpl(this.api, ci);
            }
        }
        else {
            return this.ordered(stream, name);
        }
    }
    getConsumerFromInfo(ci) {
        if (typeof ci.config.deliver_subject === "string") {
            throw new Error("not a pull consumer");
        }
        return new consumer_1.PullConsumerImpl(this.api, ci);
    }
    async ordered(stream, opts = {}) {
        await this.checkVersion();
        const impl = this.api;
        const sapi = new StreamAPIImpl(impl.nc, impl.opts);
        await sapi.info(stream);
        if (typeof opts.name_prefix === "string") {
            (0, jsutil_1.minValidation)("name_prefix", opts.name_prefix);
        }
        opts.name_prefix = opts.name_prefix || internal_1.nuid.next();
        const name = `${opts.name_prefix}_1`;
        const config = {
            name,
            deliver_policy: jsapi_types_1.DeliverPolicy.StartSequence,
            opt_start_seq: opts.opt_start_seq || 1,
            ack_policy: jsapi_types_1.AckPolicy.None,
            inactive_threshold: (0, internal_1.nanos)(5 * 60 * 1000),
            num_replicas: 1,
            max_deliver: 1,
            mem_storage: true,
        };
        if (opts.headers_only === true) {
            config.headers_only = true;
        }
        if (Array.isArray(opts.filter_subjects)) {
            config.filter_subjects = opts.filter_subjects;
        }
        if (typeof opts.filter_subjects === "string") {
            config.filter_subject = opts.filter_subjects;
        }
        if (opts.replay_policy) {
            config.replay_policy = opts.replay_policy;
        }
        config.deliver_policy = opts.deliver_policy ||
            jsapi_types_1.DeliverPolicy.StartSequence;
        if (opts.deliver_policy === jsapi_types_1.DeliverPolicy.All ||
            opts.deliver_policy === jsapi_types_1.DeliverPolicy.LastPerSubject ||
            opts.deliver_policy === jsapi_types_1.DeliverPolicy.New ||
            opts.deliver_policy === jsapi_types_1.DeliverPolicy.Last) {
            delete config.opt_start_seq;
            config.deliver_policy = opts.deliver_policy;
        }
        // this requires a filter subject - we only set if they didn't
        // set anything, and to be pre-2.10 we set it as filter_subject
        if (config.deliver_policy === jsapi_types_1.DeliverPolicy.LastPerSubject) {
            if (typeof config.filter_subjects === "undefined" &&
                typeof config.filter_subject === "undefined") {
                config.filter_subject = ">";
            }
        }
        if (opts.opt_start_time) {
            delete config.opt_start_seq;
            config.deliver_policy = jsapi_types_1.DeliverPolicy.StartTime;
            config.opt_start_time = opts.opt_start_time;
        }
        if (opts.inactive_threshold) {
            config.inactive_threshold = (0, internal_1.nanos)(opts.inactive_threshold);
        }
        const ci = await this.api.add(stream, config);
        return Promise.resolve(new consumer_1.PullConsumerImpl(this.api, ci, opts));
    }
}
exports.ConsumersImpl = ConsumersImpl;
class StreamImpl {
    api;
    _info;
    constructor(api, info) {
        this.api = api;
        this._info = info;
    }
    get name() {
        return this._info.config.name;
    }
    alternates() {
        return this.info()
            .then((si) => {
            return si.alternates ? si.alternates : [];
        });
    }
    async best() {
        await this.info();
        if (this._info.alternates) {
            const asi = await this.api.info(this._info.alternates[0].name);
            return new StreamImpl(this.api, asi);
        }
        else {
            return this;
        }
    }
    info(cached = false, opts) {
        if (cached) {
            return Promise.resolve(this._info);
        }
        return this.api.info(this.name, opts)
            .then((si) => {
            this._info = si;
            return this._info;
        });
    }
    getConsumer(name) {
        return new ConsumersImpl(new jsmconsumer_api_1.ConsumerAPIImpl(this.api.nc, this.api.opts))
            .get(this.name, name);
    }
    getPushConsumer(name) {
        return new ConsumersImpl(new jsmconsumer_api_1.ConsumerAPIImpl(this.api.nc, this.api.opts))
            .getPushConsumer(this.name, name);
    }
    getMessage(query) {
        return this.api.getMessage(this.name, query);
    }
    deleteMessage(seq, erase) {
        return this.api.deleteMessage(this.name, seq, erase);
    }
}
exports.StreamImpl = StreamImpl;
class StreamAPIImpl extends jsbaseclient_api_1.BaseApiClientImpl {
    constructor(nc, opts) {
        super(nc, opts);
    }
    checkStreamConfigVersions(cfg) {
        const nci = this.nc;
        if (cfg.metadata) {
            const { min, ok } = nci.features.get(internal_1.Feature.JS_STREAM_CONSUMER_METADATA);
            if (!ok) {
                throw new Error(`stream 'metadata' requires server ${min}`);
            }
        }
        if (cfg.first_seq) {
            const { min, ok } = nci.features.get(internal_1.Feature.JS_STREAM_FIRST_SEQ);
            if (!ok) {
                throw new Error(`stream 'first_seq' requires server ${min}`);
            }
        }
        if (cfg.subject_transform) {
            const { min, ok } = nci.features.get(internal_1.Feature.JS_STREAM_SUBJECT_TRANSFORM);
            if (!ok) {
                throw new Error(`stream 'subject_transform' requires server ${min}`);
            }
        }
        if (cfg.compression) {
            const { min, ok } = nci.features.get(internal_1.Feature.JS_STREAM_COMPRESSION);
            if (!ok) {
                throw new Error(`stream 'compression' requires server ${min}`);
            }
        }
        if (cfg.consumer_limits) {
            const { min, ok } = nci.features.get(internal_1.Feature.JS_DEFAULT_CONSUMER_LIMITS);
            if (!ok) {
                throw new Error(`stream 'consumer_limits' requires server ${min}`);
            }
        }
        function validateStreamSource(context, src) {
            const count = src?.subject_transforms?.length || 0;
            if (count > 0) {
                const { min, ok } = nci.features.get(internal_1.Feature.JS_STREAM_SOURCE_SUBJECT_TRANSFORM);
                if (!ok) {
                    throw new Error(`${context} 'subject_transforms' requires server ${min}`);
                }
            }
        }
        if (cfg.sources) {
            cfg.sources.forEach((src) => {
                validateStreamSource("stream sources", src);
            });
        }
        if (cfg.mirror) {
            validateStreamSource("stream mirror", cfg.mirror);
        }
    }
    async add(cfg) {
        this.checkStreamConfigVersions(cfg);
        (0, jsutil_1.validateStreamName)(cfg.name);
        cfg.mirror = convertStreamSourceDomain(cfg.mirror);
        //@ts-ignore: the sources are either set or not - so no item should be undefined in the list
        cfg.sources = cfg.sources?.map(convertStreamSourceDomain);
        const r = await this._request(`${this.prefix}.STREAM.CREATE.${cfg.name}`, cfg);
        const si = r;
        this._fixInfo(si);
        return si;
    }
    async delete(stream) {
        (0, jsutil_1.validateStreamName)(stream);
        const r = await this._request(`${this.prefix}.STREAM.DELETE.${stream}`);
        const cr = r;
        return cr.success;
    }
    async update(name, cfg = {}) {
        if (typeof name === "object") {
            const sc = name;
            name = sc.name;
            cfg = sc;
            console.trace(`\u001B[33m >> streams.update(config: StreamConfig) api changed to streams.update(name: string, config: StreamUpdateConfig) - this shim will be removed - update your code.  \u001B[0m`);
        }
        this.checkStreamConfigVersions(cfg);
        (0, jsutil_1.validateStreamName)(name);
        const old = await this.info(name);
        const update = Object.assign(old.config, cfg);
        update.mirror = convertStreamSourceDomain(update.mirror);
        //@ts-ignore: the sources are either set or not - so no item should be undefined in the list
        update.sources = update.sources?.map(convertStreamSourceDomain);
        const r = await this._request(`${this.prefix}.STREAM.UPDATE.${name}`, update);
        const si = r;
        this._fixInfo(si);
        return si;
    }
    async info(name, data) {
        (0, jsutil_1.validateStreamName)(name);
        const subj = `${this.prefix}.STREAM.INFO.${name}`;
        const r = await this._request(subj, data);
        let si = r;
        let { total, limit } = si;
        // check how many subjects we got in the first request
        let have = si.state.subjects
            ? Object.getOwnPropertyNames(si.state.subjects).length
            : 1;
        // if the response is paged, we have a large list of subjects
        // handle the paging and return a StreamInfo with all of it
        if (total && total > have) {
            const infos = [si];
            const paged = data || {};
            let i = 0;
            // total could change, so it is possible to have collected
            // more that the total
            while (total > have) {
                i++;
                paged.offset = limit * i;
                const r = await this._request(subj, paged);
                // update it in case it changed
                total = r.total;
                infos.push(r);
                const count = Object.getOwnPropertyNames(r.state.subjects).length;
                have += count;
                // if request returns less than limit it is done
                if (count < limit) {
                    // done
                    break;
                }
            }
            // collect all the subjects
            let subjects = {};
            for (let i = 0; i < infos.length; i++) {
                si = infos[i];
                if (si.state.subjects) {
                    subjects = Object.assign(subjects, si.state.subjects);
                }
            }
            // don't give the impression we paged
            si.offset = 0;
            si.total = 0;
            si.limit = 0;
            si.state.subjects = subjects;
        }
        this._fixInfo(si);
        return si;
    }
    list(subject = "") {
        const payload = subject?.length ? { subject } : {};
        const listerFilter = (v) => {
            const slr = v;
            slr.streams.forEach((si) => {
                this._fixInfo(si);
            });
            return slr.streams;
        };
        const subj = `${this.prefix}.STREAM.LIST`;
        return new jslister_1.ListerImpl(subj, listerFilter, this, payload);
    }
    // FIXME: init of sealed, deny_delete, deny_purge shouldn't be necessary
    //  https://github.com/nats-io/nats-server/issues/2633
    _fixInfo(si) {
        si.config.sealed = si.config.sealed || false;
        si.config.deny_delete = si.config.deny_delete || false;
        si.config.deny_purge = si.config.deny_purge || false;
        si.config.allow_rollup_hdrs = si.config.allow_rollup_hdrs || false;
    }
    async purge(name, opts) {
        if (opts) {
            const { keep, seq } = opts;
            if (typeof keep === "number" && typeof seq === "number") {
                throw internal_1.InvalidArgumentError.format(["keep", "seq"], "are mutually exclusive");
            }
        }
        (0, jsutil_1.validateStreamName)(name);
        const v = await this._request(`${this.prefix}.STREAM.PURGE.${name}`, opts);
        return v;
    }
    async deleteMessage(stream, seq, erase = true) {
        (0, jsutil_1.validateStreamName)(stream);
        const dr = { seq };
        if (!erase) {
            dr.no_erase = true;
        }
        const r = await this._request(`${this.prefix}.STREAM.MSG.DELETE.${stream}`, dr);
        const cr = r;
        return cr.success;
    }
    async getMessage(stream, query) {
        (0, jsutil_1.validateStreamName)(stream);
        try {
            const r = await this._request(`${this.prefix}.STREAM.MSG.GET.${stream}`, query);
            const sm = r;
            return new StoredMsgImpl(sm);
        }
        catch (err) {
            if (err instanceof jserrors_1.JetStreamApiError &&
                err.code === jserrors_1.JetStreamApiCodes.NoMessageFound) {
                return null;
            }
            return Promise.reject(err);
        }
    }
    find(subject) {
        return this.findStream(subject);
    }
    names(subject = "") {
        const payload = subject?.length ? { subject } : {};
        const listerFilter = (v) => {
            const sr = v;
            return sr.streams;
        };
        const subj = `${this.prefix}.STREAM.NAMES`;
        return new jslister_1.ListerImpl(subj, listerFilter, this, payload);
    }
    async get(name) {
        const si = await this.info(name);
        return Promise.resolve(new StreamImpl(this, si));
    }
}
exports.StreamAPIImpl = StreamAPIImpl;
class StoredMsgImpl {
    _header;
    smr;
    static jc;
    constructor(smr) {
        this.smr = smr;
    }
    get pending() {
        return 0;
    }
    get lastSequence() {
        return 0;
    }
    get subject() {
        return this.smr.message.subject;
    }
    get seq() {
        return this.smr.message.seq;
    }
    get timestamp() {
        return this.smr.message.time;
    }
    get time() {
        return new Date(Date.parse(this.timestamp));
    }
    get data() {
        return this.smr.message.data ? this._parse(this.smr.message.data) : internal_1.Empty;
    }
    get header() {
        if (!this._header) {
            if (this.smr.message.hdrs) {
                const hd = this._parse(this.smr.message.hdrs);
                this._header = internal_1.MsgHdrsImpl.decode(hd);
            }
            else {
                this._header = (0, internal_1.headers)();
            }
        }
        return this._header;
    }
    _parse(s) {
        const bs = atob(s);
        const len = bs.length;
        const bytes = new Uint8Array(len);
        for (let i = 0; i < len; i++) {
            bytes[i] = bs.charCodeAt(i);
        }
        return bytes;
    }
    json(reviver) {
        return JSON.parse(new TextDecoder().decode(this.data), reviver);
    }
    string() {
        return internal_1.TD.decode(this.data);
    }
}
exports.StoredMsgImpl = StoredMsgImpl;
class StreamsImpl {
    api;
    constructor(api) {
        this.api = api;
    }
    get(stream) {
        return this.api.info(stream)
            .then((si) => {
            return new StreamImpl(this.api, si);
        });
    }
}
exports.StreamsImpl = StreamsImpl;
