// @nats-io/jetstream/lib/pushconsumer.js
"use strict";
/*
 * Copyright 2024 Synadia Communications, Inc
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.PushConsumerImpl = exports.PushConsumerMessagesImpl = void 0;
const jsmsg_1 = require("./jsmsg");
const jsapi_types_1 = require("./jsapi_types");
const internal_1 = require("@nats-io/nats-core/internal");
const jserrors_1 = require("./jserrors");
class PushConsumerMessagesImpl extends internal_1.QueuedIteratorImpl {
    consumer;
    sub;
    monitor;
    listeners;
    abortOnMissingResource;
    callback;
    ordered;
    cursor;
    namePrefix;
    deliverPrefix;
    serial;
    createFails;
    statusIterator;
    cancelables;
    constructor(c, userOptions = {}, internalOptions = {}) {
        super();
        this.consumer = c;
        this.monitor = null;
        this.listeners = [];
        this.cancelables = [];
        this.abortOnMissingResource =
            userOptions.abort_on_missing_resource === true;
        this.callback = userOptions.callback || null;
        this.noIterator = this.callback !== null;
        this.namePrefix = null;
        this.deliverPrefix = null;
        this.ordered = internalOptions.ordered === true;
        this.serial = 1;
        if (this.ordered) {
            this.namePrefix = internalOptions.name_prefix ?? `oc_${internal_1.nuid.next()}`;
            // this already should be set
            this.deliverPrefix = internalOptions.deliver_prefix ??
                (0, internal_1.createInbox)(this.consumer.api.nc.options.inboxPrefix);
            this.cursor = { stream_seq: 1, deliver_seq: 0 };
            const startSeq = c._info.config.opt_start_seq || 0;
            this.cursor.stream_seq = startSeq > 0 ? startSeq - 1 : 0;
            this.createFails = 0;
        }
        this.start();
    }
    reset() {
        const { name } = this.consumer._info?.config;
        if (name) {
            this.consumer.api.delete(this.consumer.stream, name)
                .catch(() => {
                // ignored
            });
        }
        const config = this.getConsumerOpts();
        // reset delivery seq
        this.cursor.deliver_seq = 0;
        // if they do a consumer info, they get the new one.
        this.consumer.name = config.name;
        // sync the serial - if they stop and restart, it will go forward
        this.consumer.serial = this.serial;
        // remap the subscription
        this.consumer.api.nc._resub(this.sub, config.deliver_subject);
        // create the consumer
        this.consumer.api.add(this.consumer.stream, config).then((ci) => {
            this.createFails = 0;
            this.consumer._info = ci;
            this.notify({ type: "ordered_consumer_recreated", name: ci.name });
        }).catch((err) => {
            this.createFails++;
            if (err.message === "stream not found") {
                this.notify({
                    type: "stream_not_found",
                    name: this.consumer.stream,
                    consumerCreateFails: this.createFails,
                });
                if (this.abortOnMissingResource) {
                    this.stop(err);
                    return;
                }
            }
            // we have attempted to create 30 times, never succeeded
            if (this.createFails >= 30 && this.received === 0) {
                this.stop(err);
            }
            const bo = (0, internal_1.backoff)();
            const c = (0, internal_1.delay)(bo.backoff(this.createFails));
            c.then(() => {
                if (!this.done) {
                    this.reset();
                }
            }).catch(() => { })
                .finally(() => {
                const idx = this.cancelables.indexOf(c);
                if (idx !== -1) {
                    this.cancelables = this.cancelables.splice(idx, idx);
                }
            });
            this.cancelables.push(c);
        });
    }
    getConsumerOpts() {
        const src = Object.assign({}, this.consumer._info.config);
        this.serial++;
        const name = `${this.namePrefix}_${this.serial}`;
        return Object.assign(src, {
            name,
            deliver_policy: jsapi_types_1.DeliverPolicy.StartSequence,
            opt_start_seq: this.cursor.stream_seq + 1,
            ack_policy: jsapi_types_1.AckPolicy.None,
            inactive_threshold: (0, internal_1.nanos)(5 * 60 * 1000),
            num_replicas: 1,
            flow_control: true,
            idle_heartbeat: (0, internal_1.nanos)(30 * 1000),
            deliver_subject: `${this.deliverPrefix}.${this.serial}`,
        });
    }
    closed() {
        return this.iterClosed;
    }
    close() {
        this.stop();
        return this.iterClosed;
    }
    stop(err) {
        if (this.done) {
            return;
        }
        this.statusIterator?.stop();
        this.monitor?.cancel();
        this.monitor = null;
        // if we have delays, stop them
        this.cancelables.forEach((c) => {
            c.cancel();
        });
        Promise.all(this.cancelables)
            .then(() => {
            this.cancelables = [];
        })
            .catch(() => { })
            .finally(() => {
            this._push(() => {
                super.stop(err);
                this.listeners.forEach((n) => {
                    n.stop();
                });
            });
        });
    }
    _push(r) {
        if (!this.callback) {
            super.push(r);
        }
        else {
            const fn = typeof r === "function" ? r : null;
            try {
                if (!fn) {
                    const m = r;
                    this.received++;
                    this.callback(m);
                    this.processed++;
                }
                else {
                    fn();
                }
            }
            catch (err) {
                this.stop(err);
            }
        }
    }
    status() {
        const iter = new internal_1.QueuedIteratorImpl();
        this.listeners.push(iter);
        return iter;
    }
    start() {
        const { deliver_subject: subject, deliver_group: queue, idle_heartbeat: hbNanos, } = this.consumer._info.config;
        if (!subject) {
            // this shouldn't happen - the push consumer should be validated
            throw new Error("bad consumer info");
        }
        if (hbNanos) {
            const ms = (0, internal_1.millis)(hbNanos);
            this.monitor = new internal_1.IdleHeartbeatMonitor(ms, (count) => {
                this.notify({ type: "heartbeats_missed", count });
                if (this.ordered) {
                    this.reset();
                }
                return false;
            }, { maxOut: 2 });
            (async () => {
                this.statusIterator = this.consumer.api.nc
                    .status();
                for await (const s of this.statusIterator) {
                    switch (s.type) {
                        case "disconnect":
                            this.monitor?.cancel();
                            break;
                        case "reconnect":
                            this.monitor?.restart();
                            break;
                        default:
                        // ignored
                    }
                }
            })();
        }
        this.sub = this.consumer.api.nc.subscribe(subject, {
            queue,
            callback: (err, msg) => {
                if (err) {
                    this.stop(err);
                    return;
                }
                this.monitor?.work();
                // need to make sure to catch all protocol messages even
                const isProtocol = this.ordered
                    ? msg.subject.indexOf(this?.deliverPrefix) === 0
                    : msg.subject === subject;
                if (isProtocol) {
                    if (msg.subject !== this.sub.getSubject()) {
                        // this is a stale message - was not sent to the current inbox
                        return;
                    }
                    const status = new jserrors_1.JetStreamStatus(msg);
                    if (status.isFlowControlRequest()) {
                        this._push(() => {
                            msg.respond();
                            this.notify({ type: "flow_control" });
                        });
                        return;
                    }
                    if (status.isIdleHeartbeat()) {
                        const lastConsumerSequence = parseInt(msg.headers?.get("Nats-Last-Consumer") || "0");
                        const lastStreamSequence = parseInt(msg.headers?.get("Nats-Last-Stream") ?? "0");
                        this.notify({
                            type: "heartbeat",
                            lastStreamSequence,
                            lastConsumerSequence,
                        });
                        return;
                    }
                    const code = status.code;
                    const description = status.description;
                    if (status.isConsumerDeleted()) {
                        this.notify({ type: "consumer_deleted", code, description });
                    }
                    if (this.abortOnMissingResource) {
                        this._push(() => {
                            this.stop(status.toError());
                        });
                        return;
                    }
                }
                else {
                    const m = (0, jsmsg_1.toJsMsg)(msg);
                    if (this.ordered) {
                        const dseq = m.info.deliverySequence;
                        if (dseq !== this.cursor.deliver_seq + 1) {
                            this.reset();
                            return;
                        }
                        this.cursor.deliver_seq = dseq;
                        this.cursor.stream_seq = m.info.streamSequence;
                    }
                    this._push(m);
                }
            },
        });
        this.sub.closed.then(() => {
            // for ordered consumer we cannot break the iterator
            this._push(() => {
                this.stop();
            });
        });
        this.closed().then(() => {
            this.sub?.unsubscribe();
        });
    }
    notify(n) {
        if (this.listeners.length > 0) {
            (() => {
                this.listeners.forEach((l) => {
                    const qi = l;
                    if (!qi.done) {
                        qi.push(n);
                    }
                });
            })();
        }
    }
}
exports.PushConsumerMessagesImpl = PushConsumerMessagesImpl;
class PushConsumerImpl {
    api;
    _info;
    stream;
    name;
    bound;
    ordered;
    started;
    serial;
    opts;
    constructor(api, info, opts = {}) {
        this.api = api;
        this._info = info;
        this.stream = info.stream_name;
        this.name = info.name;
        this.bound = opts.bound === true;
        this.started = false;
        this.opts = opts;
        this.serial = 0;
        this.ordered = opts.ordered || false;
        if (this.ordered) {
            this.serial = 1;
        }
    }
    consume(userOptions = {}) {
        userOptions = { ...userOptions };
        if (this.started) {
            return Promise.reject(new internal_1.errors.InvalidOperationError("consumer already started"));
        }
        if (!this._info.config.deliver_subject) {
            return Promise.reject(new Error("deliver_subject is not set, not a push consumer"));
        }
        if (!this._info.config.deliver_group && this._info.push_bound) {
            return Promise.reject(new internal_1.errors.InvalidOperationError("consumer is already bound"));
        }
        const v = new PushConsumerMessagesImpl(this, userOptions, this.opts);
        this.started = true;
        v.closed().then(() => {
            this.started = false;
        });
        return Promise.resolve(v);
    }
    delete() {
        if (this.bound) {
            return Promise.reject(new internal_1.errors.InvalidOperationError("bound consumers cannot delete"));
        }
        const { stream_name, name } = this._info;
        return this.api.delete(stream_name, name);
    }
    async info(cached) {
        if (this.bound) {
            return Promise.reject(new internal_1.errors.InvalidOperationError("bound consumers cannot info"));
        }
        if (cached) {
            return Promise.resolve(this._info);
        }
        // FIXME: this can possibly return a stale ci if this is an ordered
        //   consumer, and the consumer reset while we awaited the info...
        const info = await this.api.info(this.stream, this.name);
        this._info = info;
        return info;
    }
    isPullConsumer() {
        return false;
    }
    isPushConsumer() {
        return true;
    }
}
exports.PushConsumerImpl = PushConsumerImpl;
