// @nats-io/jetstream/lib/types.js
"use strict";
/*
 * Copyright 2023-2025 The NATS Authors
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RepublishHeaders = exports.DirectMsgHeaders = exports.JsHeaders = exports.AdvisoryKind = void 0;
exports.isOrderedPushConsumerOptions = isOrderedPushConsumerOptions;
exports.isPullConsumer = isPullConsumer;
exports.isPushConsumer = isPushConsumer;
exports.isBoundPushConsumerOptions = isBoundPushConsumerOptions;
function isOrderedPushConsumerOptions(v) {
    if (v && typeof v === "object") {
        return "name_prefix" in v ||
            "deliver_subject_prefix" in v ||
            "filter_subjects" in v ||
            "filter_subject" in v ||
            "deliver_policy" in v ||
            "opt_start_seq" in v ||
            "opt_start_time" in v ||
            "replay_policy" in v ||
            "inactive_threshold" in v ||
            "headers_only" in v ||
            "deliver_prefix" in v;
    }
    return false;
}
function isPullConsumer(v) {
    return v.isPullConsumer();
}
function isPushConsumer(v) {
    return v.isPushConsumer();
}
function isBoundPushConsumerOptions(v) {
    if (v && typeof v === "object") {
        return "deliver_subject" in v ||
            "deliver_group" in v ||
            "idle_heartbeat" in v;
    }
    return false;
}
/**
 * The different kinds of Advisories
 */
exports.AdvisoryKind = {
    API: "api_audit",
    StreamAction: "stream_action",
    ConsumerAction: "consumer_action",
    SnapshotCreate: "snapshot_create",
    SnapshotComplete: "snapshot_complete",
    RestoreCreate: "restore_create",
    RestoreComplete: "restore_complete",
    MaxDeliver: "max_deliver",
    Terminated: "terminated",
    Ack: "consumer_ack",
    StreamLeaderElected: "stream_leader_elected",
    StreamQuorumLost: "stream_quorum_lost",
    ConsumerLeaderElected: "consumer_leader_elected",
    ConsumerQuorumLost: "consumer_quorum_lost",
};
exports.JsHeaders = {
    /**
     * Set if message is from a stream source - format is `stream seq`
     */
    StreamSourceHdr: "Nats-Stream-Source",
    /**
     * Set for heartbeat messages
     */
    LastConsumerSeqHdr: "Nats-Last-Consumer",
    /**
     * Set for heartbeat messages
     */
    LastStreamSeqHdr: "Nats-Last-Stream",
    /**
     * Set for heartbeat messages if the consumer is stalled
     */
    ConsumerStalledHdr: "Nats-Consumer-Stalled",
    /**
     * Set for headers_only consumers indicates the number of bytes in the payload
     */
    MessageSizeHdr: "Nats-Msg-Size",
    // rollup header
    RollupHdr: "Nats-Rollup",
    // value for rollup header when rolling up a subject
    RollupValueSubject: "sub",
    // value for rollup header when rolling up all subjects
    RollupValueAll: "all",
    /**
     * Set on protocol messages to indicate pull request message count that
     * was not honored.
     */
    PendingMessagesHdr: "Nats-Pending-Messages",
    /**
     * Set on protocol messages to indicate pull request byte count that
     * was not honored
     */
    PendingBytesHdr: "Nats-Pending-Bytes",
};
exports.DirectMsgHeaders = {
    Stream: "Nats-Stream",
    Sequence: "Nats-Sequence",
    TimeStamp: "Nats-Time-Stamp",
    Subject: "Nats-Subject",
    LastSequence: "Nats-Last-Sequence",
    NumPending: "Nats-Num-Pending",
};
exports.RepublishHeaders = {
    /**
     * The source stream of the message
     */
    Stream: "Nats-Stream",
    /**
     * The original subject of the message
     */
    Subject: "Nats-Subject",
    /**
     * The sequence of the republished message
     */
    Sequence: "Nats-Sequence",
    /**
     * The stream sequence id of the last message ingested to the same original subject (or 0 if none or deleted)
     */
    LastSequence: "Nats-Last-Sequence",
    /**
     * The size in bytes of the message's body - Only if {@link Republish#headers_only} is set.
     */
    Size: "Nats-Msg-Size",
};
