// @nats-io/kv/lib/internal_mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateKey = exports.validateBucket = exports.NoopKvCodecs = exports.Kvm = exports.defaultBucketOpts = exports.Bucket = exports.Base64KeyCodec = exports.KvWatchInclude = exports.kvPrefix = void 0;
var types_1 = require("./types");
Object.defineProperty(exports, "kvPrefix", { enumerable: true, get: function () { return types_1.kvPrefix; } });
Object.defineProperty(exports, "KvWatchInclude", { enumerable: true, get: function () { return types_1.KvWatchInclude; } });
var kv_1 = require("./kv");
Object.defineProperty(exports, "Base64KeyCodec", { enumerable: true, get: function () { return kv_1.Base64KeyCodec; } });
Object.defineProperty(exports, "Bucket", { enumerable: true, get: function () { return kv_1.Bucket; } });
Object.defineProperty(exports, "defaultBucketOpts", { enumerable: true, get: function () { return kv_1.defaultBucketOpts; } });
Object.defineProperty(exports, "Kvm", { enumerable: true, get: function () { return kv_1.Kvm; } });
Object.defineProperty(exports, "NoopKvCodecs", { enumerable: true, get: function () { return kv_1.NoopKvCodecs; } });
Object.defineProperty(exports, "validateBucket", { enumerable: true, get: function () { return kv_1.validateBucket; } });
Object.defineProperty(exports, "validateKey", { enumerable: true, get: function () { return kv_1.validateKey; } });
